client/main.lua

```diff
@@ -555,101 +516,47 @@ local function createShowroomVehiclePoint(data)
-                        lib.showTextUI(locale('menus.keypress_showFinanceMenu'))
-                    end,
-                    inside = function()
-                        if IsControlJustPressed(0, 38) then
-                            showFinancedVehiclesMenu()
-                        end
-                    end,
-                    onExit = function()
-                        lib.hideTextUI()
-                    end
-                })
-            end
-        end
-    end)
+--- End test drive
+local function endTestDrive()
+    LocalPlayer.state:set('isInTestDrive', false, true)
+    exports.qbx_core:Notify(locale('general.testdrive_complete'), 'success')
+end
+
+--- Starts the test drive timer based on time and shop
+---@param time integer
+local function startTestDriveTimer(time)
+    local gameTimer = GetGameTimer()
+    local timeMs = time * 1000
 
     CreateThread(function()
-        for shopName, shop in pairs(sharedConfig.shops) do
-            createShop(shop.zone.shape, shopName)
-            showroomPoints[shopName] = {}
-
-            local showroomVehicles = sharedConfig.shops[shopName].showroomVehicles
-            for i = 1, #showroomVehicles do
-                local showroomVehicle = showroomVehicles[i]
-                showroomPoints[shopName][i] = createShowroomVehiclePoint({
-                    coords = showroomVehicle.coords,
-                    shopName = shopName,
-                    vehiclePos = i,
-                    model = showroomVehicle.vehicle
-                })
+        local playerState = LocalPlayer.state
+        while playerState.isInTestDrive do
+            local currentGameTime = GetGameTimer()
+            local secondsLeft = currentGameTime - gameTimer
+
+            if currentGameTime < gameTimer + timeMs and secondsLeft >= timeMs - 50 then
+                endTestDrive()
             end
+
+            qbx.drawText2d({
+                text = locale('general.testdrive_timer')..math.ceil(time - secondsLeft / 1000),
+                coords = vec2(1.0, 1.38),
+                scale = 0.5
+            })
+
+            Wait(0)
         end
     end)
 end
 
---- Executes once player fully loads in
-AddEventHandler('QBCore:Client:OnPlayerLoaded', function()
-    init()
-end)
+AddStateBagChangeHandler('isInTestDrive', ('player:%s'):format(cache.serverId), function(_, _, value)
+    if not value then return end
 
---- Starts the test drive. If vehicle parameter is not provided then the test drive will start with the closest vehicle to the player.
---- @param args table
-RegisterNetEvent('qbx_vehicleshop:client:testDrive', function(args)
-    if LocalPlayer.state.isInTestDrive then
-        exports.qbx_core:Notify(locale('error.testdrive_alreadyin'), 'error')
-        return
+    while not cache.vehicle do
+        Wait(10)
     end
 
-    if not args then return end
-
-    LocalPlayer.state:set('isInTestDrive', true, true)
     local testDrive = sharedConfig.shops[insideShop].testDrive
```

Comment:
This doesn't work because by the time the statebag is changed, the player has already been warped inside the test vehicle and is outside the shop. Thus, `insideShop` no longer exists. It's probably possible to save the last shop or change the statebag to contain the time limit.